charmcraft/models/project.py

```diff
@@ -0,0 +1,480 @@
+    craft-application rather than the legacy charmcraft entrypoint. Eventually
+    it will be the only form of the project.
+
+    This inherits from CraftBaseModel rather than from the base craft-application Project
+    in order to preserve field order. It's registered as a virtual child class below.
+    """
+
+    type: Literal["charm", "bundle"]
+    name: Optional[models.ProjectName]
+    title: Optional[models.ProjectTitle]
+    summary: Optional[models.SummaryStr]
+    description: Optional[str]
+
+    analysis: Optional[AnalysisConfig]
+    charmhub: Optional[CharmhubConfig]
+    parts: Optional[Dict[str, Dict[str, Any]]]  # parts are handled by craft-parts
+
+    # Default project properties that Charmcraft currently does not use. Types are set
+    # to be Optional[None], preventing them from being used, but allow them to be used
+    # by the application.
+    version: Optional[None] = None
+    base: Optional[None] = None
+    license: Optional[None] = None
+    # These are inside the "links" child model.
+    contact: Optional[None] = None
+    issues: Optional[None] = None
+    source_code: Optional[None] = None
+
+    # These private attributes are not part of the project model but are attached here
+    # because Charmcraft uses this metadata.
+    _started_at: datetime.datetime = pydantic.PrivateAttr(default_factory=datetime.datetime.utcnow)
+    _valid: bool = pydantic.PrivateAttr(default=False)
+
+    @property
+    def started_at(self) -> datetime.datetime:
+        """Get the time that Charmcraft started running."""
+        return self._started_at
+
+    @classmethod
+    def unmarshal(cls, data: Dict[str, Any]):
+        """Create a Charmcraft project from a dictionary of data."""
+        project_type = data.get("type")
+        if project_type == "charm":
+            return Charm.parse_obj(data)
+        if project_type == "bundle":
+            return Bundle.parse_obj(data)
+        raise ValueError(f"field type cannot be {project_type!r}")
+
+    @classmethod
+    def from_yaml_file(cls, path: pathlib.Path) -> Self:
+        """Instantiate this model from a YAML file.
+
+        For use with craft-application.
+        """
+        try:
+            with path.open() as file:
+                data = safe_yaml_load(file)
+        except OSError as exc:
+            raise CraftError(
+                f"Error parsing charmcraft.yaml at '{path}'", details=exc.strerror
+            ) from exc
+
+        if not isinstance(data, dict):
+            raise errors.CraftValidationError(
+                "Invalid 'charmcraft.yaml' file",
+                details=f"File generated a {type(data)} object, expected a dictionary",
+                resolution="Ensure 'charmcraft.yaml' is valid",
+                reportable=False,
+                docs_url="https://juju.is/docs/sdk/charmcraft-yaml",
+            )
+
+        charm_dir = path.parent
+
+        metadata_file = charm_dir / METADATA_FILENAME
```

Comment:
I suppose in the future this code could go in https://github.com/canonical/craft-application/blob/main/craft_application/application.py#L415